### PR TITLE
[1458] Switch factories to use the JSONAPI renderer

### PR DESCRIPTION
**Pull request by heidar**
### Context

Using the `jsonapi-rb` gems to render JSONAPI responses based on our factories. 

### Changes proposed in this pull request

A few lines of metaprogramming magic to create serializers and make our models able to turn themselves into JSONAPI hashes.

### Guidance to review

TODO:

* Still need to do this for the rest of the factories (should be easy)
* Relations (easy to add in)
* Custom meta attributes (might be easier to declare the serializers manually for these)
* Need to support the `includes` parameter (should be easy)
* Need to figure out how to make this work with `build_list` (a bit of a pain)

### CODE REVIEW

File: spec/support/initialize_factory_serializers.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FactoryBot.factories.each do |factory|
+  next unless factory.name.in? %i[user]
+
+  serializer_klass = "#{factory.name.capitalize}Serializer"
+
+  Object.const_set(
```
Comment (defong) [inline spec/support/initialize_factory_serializers.rb]: Magic code :heart_eyes:

Comment (misaka) [reply spec/support/initialize_factory_serializers.rb]: Ya, can we document what's happening here please?

Comment (misaka) [reply spec/support/initialize_factory_serializers.rb]: I _think_ that you're defining a new class on-the-fly here ... is there a clearer way to do it? Even moving the `Class.new` out of the `Object.const_set` might help imo.

File: spec/support/define_factory_jsonapi_render_methods.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+FactoryBot.define do
+  after :build do |record|
```
Comment (misaka) [inline spec/support/define_factory_jsonapi_render_methods.rb]: This little gem could also use a little explanation.

File: spec/support/initialize_factory_serializers.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+FactoryBot.factories.each do |factory|
+  next unless factory.name.in? %i[user]
+
+  # Create a string that holds the name of the class, for example
+  # "UserSerializer"
+  serializer_class_name = "#{factory.name.capitalize}Serializer"
+
+  # Create a new class (of type Class) and make it inherit from the JSONAPI
+  # serializer resource. Everything in the block is realted to the serializer.
```
Comment (defong) [inline spec/support/initialize_factory_serializers.rb]: realted => related

```suggestion
  # serializer resource. Everything in the block is related to the serializer.
```

File: spec/support/initialize_factory_serializers.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  # "users".
+  # We set the attributes to render out to the attributes we define in the
+  # factory.
+  serializer_class      = Class.new(JSONAPI::Serializable::Resource) do
+    type(factory.name.to_s.pluralize)
+    attributes(*FactoryBot.attributes_for(factory.name).keys)
+  end
+
+  # Create a new constant with the class name and make it point to the class
+  # we created. For example UserSerializer, and it would point to the class we
```
Comment (defong) [inline spec/support/initialize_factory_serializers.rb]:
```suggestion
  # we created. For example "UserSerializer", and it would point to the class we
```

**Comment by heidar** [comment]
[Fix tests in mc-fe](https://trello.com/c/wBdaVsDJ/1458-fix-tests-in-mc-fe)

**Pull request merged**